### PR TITLE
Fix handling of shared-line comments

### LINKED ISSUE #2237: Change *-empty-line-before keywords behaviour to ignore shared-line comments
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

This issue is similar to #2212. I want to normalize behaviour with shared-line comments and ignore them for more options. If previous node is shared-line comment, than we should check node before shared-line comment.

> Which rule, if any, is this issue related to?

`at-rule-empty-line-before`
except: ["blockless-after-same-name-blockless", "blockless-group"]
ignore: ["blockless-after-same-name-blockless"] ("blockless-group" for `ignore` has very different criteria than for `except`. #2240)

`custom-property-empty-line-before`
except: ["after-custom-property"]

`declaration-empty-line-before`
except: ["after-declaration"]
ignore: ["after-declaration"]

`rule-nested-empty-line-before`
except: ["after-rule"]

> What CSS is needed to reproduce this issue?

```css
a {

    color: #0098db; /* sfs */
    text-decoration: none;
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "declaration-empty-line-before": ["always", {
        "except": ["after-declaration"]
    }],
  }
}
```

> Which version of stylelint are you using?

e.g. `7.7.1`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No.

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors you are getting)?

The following warning was flagged:

```
4:5 error Expected empty line before declaration (declaration-empty-line-before)
```

### LINKED ISSUE #2239: Standardise checks for ignore/except "after-declaration" in declaration-empty-line-before
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

[`except: ["after-declaration"]`](https://github.com/stylelint/stylelint/blob/master/lib/rules/declaration-empty-line-before/index.js#L106-L115) checks previous node more accurately than [`ignore: ["after-declaration"]`](https://github.com/stylelint/stylelint/blob/master/lib/rules/declaration-empty-line-before/index.js#L68-L75). I think `ignore` should check same as `except`.

> Which rule, if any, is this issue related to?

`declaration-empty-line-before`

> Which version of stylelint are you using?

`7.7.1`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No.

### LINKED ISSUE #2240: Standardise checks for ignore/except "blockless-group" in at-rule-empty-line-before
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

[`except: ["blockless-group"]`](https://github.com/stylelint/stylelint/blob/master/lib/rules/at-rule-empty-line-before/index.js#L107-L108) checks previous node more accurately than [`ignore: ["blockless-group"]`](https://github.com/stylelint/stylelint/blob/master/lib/rules/at-rule-empty-line-before/index.js#L62-L63). I think `ignore` should check same as `except`.

> Which rule, if any, is this issue related to?

`at-rule-empty-line-before`

> Which version of stylelint are you using?

`7.7.1`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No.

### PR DESCRIPTION
**Pull request by davidtheclark**
Introduces a number of utilities and adjusts the following rules: `at-rule-empty-line-before`, `custom-property-empty-line-before`, `declaration-empty-line-before`, `rule-nested-empty-line-before`, `rule-non-nested-empty-line-before`.

Closes #2237; closes #2239; closes #2240.

This turned out to be a lot of work 😅 

cc @jeddy3 @hudochenkov

### CODE REVIEW

**Comment by jeddy3** [comment]
> This turned out to be a lot of work 😅

I bet! It's looking great though.

However, (and I don't think you're going to like this as it might mean there's a little more legwork to do), there could be a [couple of conflicts](https://github.com/stylelint/stylelint/issues/2237#issuecomment-271158968) down the line if we merge this now.  In https://github.com/stylelint/stylelint/pull/2213 we standardised a couple of options. This PR was merged into the `deprecations` branch. We're waiting on https://github.com/stylelint/stylelint/issues/2123 before merging into `master` (so that all the deprecations go out in one go). There is now nothing blocking https://github.com/stylelint/stylelint/issues/2123 btw as the deprecation list has been finalised with the closure of https://github.com/stylelint/stylelint/issues/2126.

I think only a small fraction of the PR is going to conflict; perhaps just the doc changes and `first-nested` stuff.

What do you think is the best way to go about merging? Originally, I thought `deprecations` into `master`, then rebase `v8` on `master`, then make these changes.

Can we merge this PR into `v8` now, and the do `deprecations` after? Or is that going to messy?

File: lib/utils/isFirstNestedStatement.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+/* @flow */
+"use strict"
+
+module.exports = function (statement/*: postcss$rule | postcss$atRule*/)/*: boolean*/ {
+  if (statement.type !== "rule" && statement.type !== "atrule") {
```
Comment (hudochenkov) [inline lib/utils/isFirstNestedStatement.js]: Not only `at-rule-empty-line-before`, `rule-nested-empty-line-before` have `first-nested` option. `comment-empty-line-before`, `custom-property-empty-line-before`, `declaration-empty-line-before` have it also. This helper can be used for more rules.

File: lib/utils/getNextNonSharedLineCommentNode.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
+/* @flow */
+"use strict"
+
+const _ = require("lodash")
+
+function getNodeLine(node/*:: ?: postcss$node*/)/*: number | void*/ {
+  return _.get(node, "source.start.line")
+}
+
+module.exports = function getNextNonSharedLineCommentNode(node/*:: ?: postcss$node*/)/*: postcss$node | void*/ {
```
Comment (hudochenkov) [inline lib/utils/getNextNonSharedLineCommentNode.js]: I don't know if it matters, but `getPreviousNonSharedLineCommentNode` has slightly different formatting for this line.

**Comment by davidtheclark** [comment]
> What do you think is the best way to go about merging?

@jeddy3 I say we merge now, fix conflicts as they arise. Fixing conflicts is always messy but inevitable ¯\_(ツ)_/¯

File: lib/rules/checkRuleEmptyLineBefore.js [hunk 3]
```diff
@@ -19,12 +31,12 @@ module.exports = function (opts) {
...
 
   // Optionally reverse the expectation for the first nested node
-  if (optionsMatches(opts.options, "except", "first-nested") && opts.rule === opts.rule.parent.first) {
+  if (optionsMatches(opts.options, "except", "first-nested") && isFirstNestedStatement(opts.rule)) {
     expectEmptyLineBefore = !expectEmptyLineBefore
   }
 
   // Optionally reverse the expectation if a rule precedes this node
-  if (optionsMatches(opts.options, "except", "after-rule") && opts.rule.prev() && opts.rule.prev().type === "rule") {
+  if (optionsMatches(opts.options, "except", "after-rule") && _.get(getPreviousNonSharedLineCommentNode(opts.rule), "type") === "rule") {
```
Comment (davidtheclark) [inline lib/rules/checkRuleEmptyLineBefore.js]: Yep, I find it more concise and readable.

File: lib/utils/__tests__/isSharedLineComment.test.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  })
+
+  it("returns true for a shared-line comment at the beginning", () => {
+    const root = postcss.parse(`
+      /* comment */ a {}
+    `)
+    expect(isSharedLineComment(root.nodes[0])).toBe(true)
+  })
+
+  it("returns true for a shared-line comment before a rule", () => {
```
Comment (davidtheclark) [inline lib/utils/__tests__/isSharedLineComment.test.js]: Thanks for checking so carefully!

**Comment by davidtheclark** [comment]
@hudochenkov Thanks for the feedback! I made the relevant fixes.

**Comment by hudochenkov** [comment]
@davidtheclark please, take a look at `comment-empty-line-before`. You can apply new helpers there.

File: lib/utils/isBlocklessAtRuleAfterBlocklessAtRule.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+const hasBlock = require("./hasBlock")
+const isSharedLineComment = require("./isSharedLineComment")
+
+module.exports = function (atRule/*: postcss$atRule*/)/*: boolean*/ {
+  if (atRule.type !== "atrule") {
+    return false
+  }
+
+  let previousNode = atRule.prev()
```
Comment (hudochenkov) [inline lib/utils/isBlocklessAtRuleAfterBlocklessAtRule.js]: What about using here `getPreviousNonSharedLineCommentNode()` instead?

File: lib/utils/isBlocklessAtRuleAfterSameNameBlocklessAtRule.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+const _ = require("lodash")
+const isBlocklessAtRuleAfterBlocklessAtRule = require("./isBlocklessAtRuleAfterBlocklessAtRule")
+const isSharedLineComment = require("./isSharedLineComment")
+
+module.exports = function (atRule/*: postcss$atRule*/)/*: boolean*/ {
+  if (!isBlocklessAtRuleAfterBlocklessAtRule(atRule)) {
+    return false
+  }
+
+  let previousNode = atRule.prev()
```
Comment (hudochenkov) [inline lib/utils/isBlocklessAtRuleAfterSameNameBlocklessAtRule.js]: What about using here `getPreviousNonSharedLineCommentNode()` instead?

File: lib/utils/isBlocklessAtRuleAfterSameNameBlocklessAtRule.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+  if (previousNode === undefined) {
+    return false
+  }
+
+  if (isSharedLineComment(previousNode)) {
+    previousNode = previousNode.prev()
+  }
+
+  return _.get(previousNode, "name") == atRule.name
```
Comment (hudochenkov) [inline lib/utils/isBlocklessAtRuleAfterSameNameBlocklessAtRule.js]: Shall we use `===` instead?

Comment (evilebottnawi) [inline lib/utils/isBlocklessAtRuleAfterSameNameBlocklessAtRule.js]: @hudochenkov what about add eslint rule to https://github.com/stylelint/eslint-config-stylelint

Comment (davidtheclark) [inline lib/utils/isBlocklessAtRuleAfterSameNameBlocklessAtRule.js]: Yes, this was an oversight.

File: lib/utils/isBlocklessAtRuleAfterBlocklessAtRule.js (hunk 5 above)
Comment (davidtheclark) [inline lib/utils/isBlocklessAtRuleAfterBlocklessAtRule.js]: Yeah — there are probably a million places this new utility could be used.

**Comment by davidtheclark** [comment]
@hudochenkov Addressed the latest few comments. Good ones. Thanks again.

**Comment by hudochenkov** [comment]
@davidtheclark I hope you don't mind I've made some changes to `comment-empty-line-before`. _#courage_

**Comment by davidtheclark** [comment]
@hudochenkov 👍  looks great, thanks. I forgot about that bit of feedback at least once, so you did the right thing.

**Comment by davidtheclark** [comment]
Closed by #2262.

**Comment by jeddy3** [comment]
Awesome job!

**Pull request merged**